streams/src/main/java/org/apache/kafka/streams/state/internals/CachingKeyValueStore.java

```diff
@@ -241,7 +240,13 @@ private void putAndMaybeForward(final ThreadCache.DirtyEntry entry,
                 }
             }
         } else {
-            wrapped().put(entry.key(), entry.newValue());
```

Comment:
This is why there was a bug under some configurations. We're leaving the wrong configuration in place while calling put on the wrapped store. We should set it to the evicted record's context before calling put. The same change is applied to the other two caching stores in this PR.